archunit-junit/junit5/engine/build.gradle

```diff
@@ -1,23 +1,29 @@
 plugins {
     id 'archunit.java-release-conventions'
+    id 'java-test-fixtures'
 }
 
 ext.moduleName = 'com.tngtech.archunit.junit5.engine'
 
-ext.minimumJavaVersion = JavaVersion.VERSION_1_8
+ext.minimumJavaVersion = JavaVersion.VERSION_1_9
 
 dependencies {
     api project(path: ':archunit')
     api project(path: ':archunit-junit5-api')
     api project(path: ':archunit-junit5-engine-api')
+
     implementation project(path: ':archunit-junit')
     dependency.addGuava { dependencyNotation, config -> implementation(dependencyNotation, config) }
     implementation dependency.slf4j
+    implementation dependency.junitPlatformLauncher
```

Comment:
We should not add any dependencies on unrelated JUnit projects. Only depend on the platform. I think this would also be a strange dependency, since the Platform Launcher launches `TestEngines`, so a dependency back from `TestEngines` to the launcher feels wrong to me :man_shrugging:

Reply:
In that case, please have a look at `AbstractTestNameFilter.findPostDiscoveryFilter`. We need a dependency on Launcher to be able to cast the discovery request to `LauncherDiscoveryRequest`. 

If you insist on removing this dependency, please suggest a way to obtain the same result. We can, for instance, use reflection all the way through, I was just trying to maintain some level of compile-time safety.

Reply:
What I don't understand is why we would need any dependencies on other libraries or tooling at all to solve this problem. The first issue is simply: "Given a system property `archunit.junit.includePattern`, only execute tests that match the pattern", right?
Say I
1) Convert the pattern to a filter for `(ArchUnit)TestDescriptor` (we could e.g. analyse the `UniqueId`, which in the end is also just a concatenation like `.../[class:some.SomeClass]/[field:someField]` that matches the info we have in the pattern very closely; just not sure how stable this is; otherwise we could add a custom property to `(Abstract)ArchUnitTestDescriptor` or sth. like that)
2) Within `com.tngtech.archunit.junit.internal.ArchUnitTestEngine.discover(..)` once all the discoveries are through, right before I return the `result`, I add something like
```
result.accept(descriptor -> {
  if (!includeFilterMatches(descriptor)) {
    descriptor.removeFromHierarchy();
  }
});
```

Wouldn't that solve the problem? Tests I could write very easily as part of `ArchUnitTestEngineTest.Discovers`:

```
// GIVEN filter is set via system property
System.setProperty("archunit.junit.includePattern", "*SomeClass*");

// WHEN classes are discovered
testEngine.discover(/* discovery request that would normally include SomeClass */);

// THEN classes matching the system property filter are excluded
Set<String> displayNames = descriptor.getChildren().stream().map(TestDescriptor::getDisplayName).collect(toSet());
assertThat(displayNames).doesNotContain("SomeClass");
```

Maybe I'm missing something, but to me it looks like this could solve the issue without any need for additional test infrastructure or dependencies on other 3rd party libraries? :thinking:

Reply:
> 1. Convert the pattern to a filter for (ArchUnit)TestDescriptor (we could e.g. analyse the UniqueId, which in the end is also just a concatenation like .../[class:some.SomeClass]/[field:someField] that matches the info we have in the pattern very closely; just not sure how stable this is; otherwise we could add a custom property to (Abstract)ArchUnitTestDescriptor or sth. like that)
2. Within com.tngtech.archunit.junit.internal.ArchUnitTestEngine.discover(..) once all the discoveries are through, right before I return the result, I add something like ```result.accept(descriptor -> {
  if (!includeFilterMatches(descriptor)) {
    descriptor.removeFromHierarchy();
  }
});```

Well, um, yes, this is essentially what this PR does. The only difference is it filters child descriptors before they are added, and not afterwards. 

> Tests I could write very easily as part of ArchUnitTestEngineTest.Discovers

Sure, I've already done that. Have a look at the relevant tests. 

> this could solve the issue without any need for additional test infrastructure or dependencies on other 3rd party libraries

Yes, absolutely. This PR does not depend on 3rd party libraries to be able to perform the filtering. It depends on 3rd party libraries solely because: 

> I committed some of the code in advance as it's kind of hard to keep 3 PRs in line, especially when one defines an abstract base class for use with the other ones :(

I can try to isolate the changes needed in the other PRs, but this will require splitting some commits. It will also take a while. Are you OK with that?

Reply:
Sorry for the delay :grimacing: I was a little occupied to do that 1.0.0-rc1 release. Yes, please, if you don't mind! Make the first PR as small and simple as possible. Also, if we don't need all that test infrastructure code yet, it's also good, because then we can delay the decision further. The smaller and simpler the change, the easier it is to understand later on when we look back at the history. Also a small code change is easy to argue for, a big one needs to be pondered really carefully, if it brings the ROI on the maintenance costs it will cause.